File: solutions/binary-tree-right-side-view/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        q.add(root);
        while (q.size() != 0) {
            int numQueue = q.size();
            List<Integer> temp = new ArrayList<>();
            for (int i = 0; i < numQueue; i++) {
                TreeNode item = q.remove();
                temp.add(item.val);
                if (item.left != null) q.add(item.left);
                if (item.right != null) q.add(item.right);
            }
            res.add(temp.get(temp.size() - 1));
        }
        System.out.println(q);
        return res;
    }
}

File: solutions/binary-tree-right-side-view/binary-tree-right-side-view.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import "fmt"

type Queue struct {
    elements []*TreeNode
    len int
}

func (q *Queue) Push(elem *TreeNode) {
    q.elements = append(q.elements, elem)
    q.len += 1
}

func (q *Queue) Pop() *TreeNode {
    item := q.elements[0]
    q.elements = q.elements[1:]
    q.len -= 1
    return item
}

func (q Queue) Len() int {
    return q.len
}

func rightSideView(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    var res []int
    q := Queue{
        elements: []*TreeNode{},
        len: 0,
    }
    q.Push(root)
    numCurrLevel := 0
    for q.Len() > 0 {
        numCurrLevel = q.Len()
        temp := []int{}
        for i:=0; i<numCurrLevel; i++ {
            t := q.Pop()
            temp = append(temp, t.Val)
            if t.Left != nil {q.Push(t.Left)}
            if t.Right != nil {q.Push(t.Right)}
        }
        res = append(res, temp[len(temp)-1])
    }
    return res
}